#!/usr/bin/perl
# OtsSEX_test_v3.pl
# by Nate Campbell
# Sex fish by GTseq
# This version of the script uses information from the version 3 genotyping output as a control for the sex marker.
# It then generates a genotypic sex and appends the sex marker to the end of the .genos file.
# The sex marker is then included in the summary file generated by the GTseq_GenoCompile_v3.pl script.
# This script uses predicted percentage of on-target reads that would be expected for the sex locus sequences if present as a control.
# Significant alteration in the GT-seq marker panel for chinook will cause the script to incorrectly report the genotypic sex.
# This script was made to work with Ots_GTseq299 panel

use strict; use warnings;

my @Files = `ls *.fastq`;
chomp ( @Files );

foreach my $samples (@Files){
	open (FILE, "<$samples") or die;
	$samples =~ s/.fastq//;
	my $genos = "$samples.genos";
	my $OT_reads = 0;
	my $primer_counts = 0;
	my $primerOT = 0;
	my $perofallOTreads = 0;
	open (READ, "<$genos") or die;
		while (<READ>) {
			if ($. == 1) {
				my @info = split ",", $_;
				my @info2 = split ":", $info[2];
				$OT_reads = $info2[1];
					}
				}
	close READ;
	open (OUT, '>>', $genos) or die "Error opening $genos\n";
	my $counts = 0;
	my $cntrl_counts = 0;
	my $sex_geno = "00";
	my $geno_class = "NA";
	while (<FILE>) {
		chomp;
		my $info_line = $_;
		my $seq_line = <FILE>;
		my $info_line2 = <FILE>;
		my $qual_line = <FILE>;
		if($seq_line =~ m/^GGTCTTGCAGTCAGGAGAGG/) {$primer_counts++}
		if(($seq_line =~ m/^GGTCTTGCAGTCAGGAGAGG/) && ($seq_line =~ m/TCAGCGAAGTGGAGAT/)) {$counts++}
	}
	if($primer_counts == 0) {$primer_counts = 1}
	$primerOT = $counts/$primer_counts*100;
	$primerOT = sprintf("%.3f", $primerOT);
	$perofallOTreads = $counts/$OT_reads*100;
	$perofallOTreads = sprintf("%.3f", $perofallOTreads);
	$cntrl_counts = $OT_reads * 0.004;
	$cntrl_counts = int ( $cntrl_counts );
	if ($cntrl_counts == 0) {$cntrl_counts = 1}
	if ($counts == 0) {$counts = 1}
	my $ratio = $cntrl_counts/$counts;
	$ratio = sprintf("%.3f", $ratio);
	if ($cntrl_counts + $counts < 10) {$sex_geno = "00", $geno_class = "NA"}
	elsif ($ratio >= 10) {$sex_geno = "XX", $geno_class = "A1HOM"}
	elsif ($ratio <= 0.1) {$sex_geno = "XY", $geno_class = "A2HOM"}
	elsif ($ratio <= 0.2) {$sex_geno = "00", $geno_class = "NA"}
	elsif ($ratio <= 5) {$sex_geno = "XY", $geno_class = "HET"}
	
	print OUT "Ots_SEXY3-1,X=$cntrl_counts,Y=$counts,$ratio,$sex_geno,$geno_class,0,0,$counts,$primerOT,$perofallOTreads\n";
close FILE;
close OUT;
}

